Srv pre fixednoise flooding dev

sararv22 opened the pull request.

dsuponitskiy commented on benchmark/src/pre_benchmarks.cpp:
```diff
@@ -0,0 +1,314 @@
+
+#include "benchmark/benchmark.h"
+
+#include "openfhe.h"
```
I have removed all includes of openfhe.h. This header is not allowed anywhere, but in the files that are in examples and extras. Also you would want to rebase your branch as it is 1 commit behind.

sararv22 replied:
rebased the branch

sararv22 commented on benchmark/src/pre_benchmarks.cpp:
```diff
@@ -0,0 +1,314 @@
+//==================================================================================
```
This can't be written like the other benchmarks because this one needs the functions to be run in a sequence to be able to benchmark. The main(), usage() or security_model cannot be removed for this reason.

davidbrucecousins replied:
Dmitry, we tried to alter this and wasted a day. It isn't a micro benchmark like any other. It is a full application example with benchmarking. If you don't want it, we will pull it and put it back in our application repo.

sararv22 commented on benchmark/src/pre_benchmarks.cpp:
```diff
@@ -0,0 +1,314 @@
+
+    const char* optstring = "m:h";
+    while ((opt = getopt_long(argc, argv, optstring, long_options, NULL)) != -1) {
+        std::cerr << "opt1: " << opt << "; optarg: " << optarg << std::endl;
```
These are benchmarks and would need to show the parameters that it is running the benchmarks for. If I have only timing numbers displayed, then it doesn't help knowing what parameters the numbers are for. There are cerr statements in other benchmark files too (poly-benchmark-64k.cpp)

davidbrucecousins replied:
see my last comment re - this is an application benchmark. Maybe it would be better for openfhe if it were in examples.

sararv22 commented on benchmark/src/pre_benchmarks.cpp:
```diff
@@ -0,0 +1,314 @@
+
+lbcrypto::CryptoContext<lbcrypto::DCRTPoly> cryptoContext;
+
+lbcrypto::EvalKey<DCRTPoly> reencryptionKey;
```
I'm using these global variables because I need the functions to be run in a sequence for the benchmarks. There are also global variables in other benchmark files (poly-benchmark-64k.cpp)

sararv22 commented on src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp:
```diff
@@ -260,6 +260,61 @@ std::pair<std::vector<NativeInteger>, uint32_t> ParameterGenerationBGVRNS::compu
+    }
+    else if (PREMode == DIVIDE_AND_ROUND_HRA) {
+        OPENFHE_THROW(config_error, "Noise Flooding not applicable for PRE DIVIDE_AND_ROUND_HRA mode");
+    }
```
INDCPA is the default and it doesn't need this functionality to set the noise flooding parameter. If I add an exception for that option, then it would fail the cases where the premode is not set and stays the default value.

sararv22 commented on benchmark/src/pre_benchmarks.cpp:
```diff
@@ -0,0 +1,314 @@
+    // without breaking decryption. Changing the plaintext modulus would require updating these other parameters for correctness.
+    lbcrypto::CCParams<lbcrypto::CryptoContextBGVRNS> parameters;
+
+    int plaintextModulus        = 2;
```
The parameters are specifically for the different security modes to allow for 10 hops

sararv22 commented on src/pke/include/scheme/cryptocontextparams-base.h:
```diff
@@ -229,6 +234,9 @@ class Params {
     void SetMaxRelinSkDeg(int maxRelinSkDeg0) {
         maxRelinSkDeg = maxRelinSkDeg0;
     }
+    void SetPREMode(ProxyReEncryptionMode premode0) {
```
I fixed this. Do we have a style guide and can you please point me to it if we have one?

yspolyakov commented:
@sararv22 Please remove the pre-benchmarks.cpp from the PR and move it to a different repository (the one that is being used for PRE applications). @davidbrucecousins can provide further clarifications.

sararv22 commented:
> @sararv22 Please remove the pre-benchmarks.cpp from the PR and move it to a different repository (the one that is being used for PRE applications). @davidbrucecousins can provide further clarifications.

removed the file and rebased the branch to latest dev

dsuponitskiy reviewed (approved):
Added in-class initialization for CryptoParametersRLWE and a new namespace NOISE_FLOODING for constant values. Also merged dev in to srv_PRE_fixednoise_flooding_dev as branch rebasing produced multiple conflicts due to the previous merge(s).

dsuponitskiy commented:
Started 4 CI jobs

dsuponitskiy commented:
All 4 CI jobs ended successfully.
